File: src/main/java/io/nthienan/phdiff/issue/IssueComparator.java
package io.nthienan.phdiff.issue;

import org.sonar.api.batch.postjob.issue.PostJobIssue;
import org.sonar.api.batch.rule.Severity;

import java.util.Comparator;
import java.util.Objects;

public class IssueComparator implements Comparator<PostJobIssue> {
    private static int compareComponentKeyAndLine(PostJobIssue left, PostJobIssue right) {
        if (!left.componentKey().equals(right.componentKey())) {
            return left.componentKey().compareTo(right.componentKey());
        }
        return compareInt(left.line(), right.line());
    }

    private static int compareSeverity(Severity leftSeverity, Severity rightSeverity) {
        if (leftSeverity.ordinal() > rightSeverity.ordinal()) {
            // Display higher severity first. Relies on Severity.ALL to be sorted by severity.
            return -1;
        } else {
            return 1;
        }
    }

    private static int compareInt(Integer leftLine, Integer rightLine) {
        if (Objects.equals(leftLine, rightLine)) {
            return 0;
        } else if (leftLine == null) {
            return -1;
        } else if (rightLine == null) {
            return 1;
        } else {
            return leftLine.compareTo(rightLine);
        }
    }

    @Override
    public int compare(PostJobIssue left, PostJobIssue right) {
        // Most severe issues should be displayed first.
        if (left == right) {
            return 0;
        }
        if (left == null) {
            return 1;
        }
        if (right == null) {
            return -1;
        }
        if (Objects.equals(left.severity(), right.severity())) {
            // When severity is the same, sort by component key to at least group issues from
            // the same file together.
            return compareComponentKeyAndLine(left, right);
        }
        return compareSeverity(left.severity(), right.severity());
    }
}

File: src/main/kotlin/io/nthienan/phdiff/issue/IssueComparator.kt
package io.nthienan.phdiff.issue

import org.sonar.api.batch.postjob.issue.PostJobIssue
import org.sonar.api.batch.rule.Severity
import java.util.Objects

/**
 * Created on 30-Jul-17.
 * @author nthienan
 */
class IssueComparator : Comparator<PostJobIssue> {

  companion object {
    fun compareInt(left: Int?, right: Int?): Int = when {
      left == right -> 0
      left == null -> -1
      right == null -> 1
      else -> left.compareTo(right)
    }

    fun compareComponentKeyAndLine(left: PostJobIssue, right: PostJobIssue): Int {
      return if (left.componentKey() != right.componentKey()) {
        left.componentKey().compareTo(right.componentKey())
      } else {
        compareInt(left.line(), right.line())
      }
    }

    fun compareSeverity(leftSeverity: Severity, rightSeverity: Severity): Int {
      // Display higher severity first. Relies on Severity.ALL to be sorted by severity.
      return if (leftSeverity.ordinal > rightSeverity.ordinal) {
        -1
      } else {
        1
      }
    }
  }

  override fun compare(left: PostJobIssue?, right: PostJobIssue?): Int = when {
    left == right -> 0
    left == null -> 1
    right == null -> -1
    Objects.equals(left.severity(), right.severity()) -> {
      // When severity is the same, sort by component key to at least group issues from
      // the same file together.
      compareComponentKeyAndLine(left, right)
    }
    else -> compareSeverity(left.severity(), right.severity())
  }
}
